lib/elaborate/elaborateReferences.ts

```diff
@@ -1,269 +1,361 @@
+    }
+  }
+
+  getProjectList(searchText: string) {
+    if (this.projectVisibilityMap === undefined) {
+      const projectParser = this.vhdlLinter.projectParser;
+      this.projectVisibilityMap = new Map();
+      this.addObjectsToMap(this.projectVisibilityMap, ...projectParser.packages);
+      for (const pkg of projectParser.packages) {
+        this.fillVisibilityMap(pkg);
       }
+      this.addObjectsToMap(this.projectVisibilityMap, ...projectParser.packageInstantiations);
+      this.addObjectsToMap(this.projectVisibilityMap, ...projectParser.contexts);
+    }
+    return this.projectVisibilityMap?.get(searchText) ?? [];
+  }
 
+  getList(reference: O.OReference, fallback = false) {
+    // find parent which is a scope
+    let key = reference.parent;
+    for (const [p] of O.scope(key)) {
+      if (I.implementsIHasDeclarations(p) || I.implementsIHasStatements(p) || p instanceof O.OPackageInstantiation || p instanceof O.OContext) {
+        key = p;
+        break;
+      }
+    }
 
-      // package names are only referable in direct visibility
-      if (directlyVisible && (object instanceof OPackage || object instanceof OPackageBody || object instanceof OEntity || object instanceof OArchitecture)) {
-        this.evaluateDefinition(reference, object, false);
+    if (!this.scopeVisibilityMap.has(key)) {
+      this.fillVisibilityMap(key);
+    }
+    const list = this.scopeVisibilityMap.get(key);
+    if (list === undefined) {
+      throw new Error('no map found');
+    }
+    const searchText = reference.referenceToken.getLText();
+    if (fallback) {
+      const fallbackList = this.fallbackVisibilityMap.get(key);
+      if (fallbackList === undefined) {
+        throw new Error('no fallback map found');
       }
+      return (list.get(searchText) ?? []).concat(fallbackList.get(searchText) ?? []);
+    } else {
+      return list.get(searchText) ?? [];
+    }
+  }
+
 
+  castToRead(reference: O.OReference) {
+    if (reference instanceof O.OWrite === false) {
+      if (reference instanceof O.OSelectedName) {
+        Object.setPrototypeOf(reference, O.OSelectedNameRead.prototype);
+      } else {
+        Object.setPrototypeOf(reference, O.ORead.prototype);
+      }
     }
   }
 
-  elaborateSelectedNames(reference: OSelectedName | OSelectedNameWrite) {
-    const [libraryToken] = reference.prefixTokens;
-    let library: OLibrary | undefined;
-    for (const [obj] of scope(reference)) {
-      if (implementsIHasLibraries(obj)) {
-        for (const findLibrary of obj.libraries) {
-          if (findLibrary.lexerToken.getLText() == libraryToken.referenceToken.getLText()) {
-            library = findLibrary;
-          }
+  link(reference: O.OReference, obj: O.ObjectBase & (I.IHasReferenceLinks | I.IHasLabel)) {
+    reference.definitions.push(obj);
+    if (I.implementsIHasLabel(obj)) {
+      obj.labelLinks.push(reference);
+    } else {
+      obj.referenceLinks.push(reference);
+    }
```

Comment:
Can there never be an object that has a label and a lexerToken?

Reply:
If so, what happens if we add that to the interface? (like IHasReferenceToken forbids lexertoken? (which is wrong))